ssas/service/tokenblacklist.go

```diff
@@ -0,0 +1,128 @@
+package service
+
+import (
+	"fmt"
+	"github.com/CMSgov/bcda-app/ssas"
+	"github.com/CMSgov/bcda-app/ssas/cfg"
+	"github.com/patrickmn/go-cache"
+	"github.com/pborman/uuid"
+	"time"
+)
+
+var (
+	TokenBlacklist Blacklist
+	// This default cache timeout value should never be used, since individual cache elements have their own timeouts
+	defaultCacheTimeout   = 24*time.Hour
+	cacheCleanupInterval  time.Duration
+	TokenCacheLifetime	  time.Duration
+	cacheRefreshFreq	  time.Duration
+	cacheRefreshTicker	  *time.Ticker
+)
+
+func init() {
+	cacheCleanupInterval = time.Duration(cfg.GetEnvInt("SSAS_TOKEN_BLACKLIST_CACHE_CLEANUP_MINUTES", 15)) * time.Minute
+	TokenCacheLifetime	 = time.Duration(cfg.GetEnvInt("SSAS_TOKEN_BLACKLIST_CACHE_TIMEOUT_MINUTES", 60*24)) * time.Minute
+	cacheRefreshFreq	 = time.Duration(cfg.GetEnvInt("SSAS_TOKEN_BLACKLIST_CACHE_REFRESH_MINUTES", 5)) * time.Minute
+	NewBlacklist(defaultCacheTimeout, cacheCleanupInterval)
+}
+
+//	NewBlacklist allows for easy Blacklist{} creation and manipulation during testing, and should not be called
+//		outside a test suite
+func NewBlacklist(cacheTimeout time.Duration, cleanupInterval time.Duration) *Blacklist {
+	// In case a Blacklist timer has already been started:
+	stopCacheRefreshTicker()
+
+	trackingID := uuid.NewRandom().String()
+	event := ssas.Event{Op: "InitBlacklist", TrackingID: trackingID}
+	ssas.OperationStarted(event)
+
+	bl := Blacklist{}
+	bl.c = cache.New(cacheTimeout, cleanupInterval)
+
+	if err := bl.LoadFromDatabase(); err != nil {
+		event.Help = "unable to load blacklist from database: " + err.Error()
+		ssas.OperationFailed(event)
```

Comment:
why is it ok to use the cache if it wasn't loaded from the db?

Reply:
Good questions.  Are you comfortable with a panic at some stage?

Reply:
Panics can be inadvertently swallowed. I think if we add an event for CacheSyncFailure (or something like that), we can use splunk alerts to monitor and respond. Let's start with this and improve with experience.